projects/dataset_recorder/data_recorder.cc

```diff
@@ -0,0 +1,246 @@
+#include<string>
+#include<vector>
+#include <thread>
+#include <mutex>
+#include <atomic>
+
+// BoB includes
+#include "common/main.h"
+#include "common/stopwatch.h"
+#include "common/background_exception_catcher.h"
+#include "common/gps.h"
+#include "common/map_coordinate.h"
+#include "common/bn055_imu.h"
+
+#include "imgproc/opencv_unwrap_360.h"
+#include "video/panoramic.h"
+
+#include <sys/stat.h>
+#include <ctime>
+
+
+
+
+
+
+// Standard C++ includes
+#include <iostream>
+
+using namespace BoBRobotics;
+using namespace BoBRobotics::ImgProc;
+using namespace BoBRobotics::Video;
+
+
+#define WIDTH 192
+#define HEIGHT 72
+
+#define IS_UNWRAP true
+#define UPDATE_INTERVAL 200
+
+
+void readCameraThreadFunc(BoBRobotics::Video::Input *cam,
+                        cv::Mat &img,
+                        std::atomic<bool> &shouldQuit,
+                        std::mutex &mutex)
+{
+
+    //auto unwrapper = cam->createUnwrapper(cv::Size(WIDTH,HEIGHT));
+
+    while(!shouldQuit) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(5));
+        std::lock_guard<std::mutex> lock(mutex);
+        cam->readFrame(img);
+
+    }
+
+
+}
+
+void saveThread(cv::Mat img, std::string fileName) {
+    cv::imwrite(fileName, img);
+}
+
+int bobMain(int argc, char* argv[])
+{
+
+    using degree_t = units::angle::degree_t;
+    std::vector<cv::Mat> images;
+    std::vector<std::string> fileNames;
+
+    // setting up gps
+    const char *path_linux = "/dev/ttyACM1"; // path for linux systems
+    BoBRobotics::GPS::Gps gps;
+    gps.connect(path_linux);
+    BoBRobotics::BN055 imu;
+    const int maxTrials = 3;
+    int numTrials = maxTrials;
+    std::atomic<bool> shouldQuit{false};
+    std::mutex mex;
+
+
+    time_t rawtime;
+    struct tm * timeinfo;
+    char buffer[80];
+
+    time (&rawtime);
+    timeinfo = localtime(&rawtime);
+
+    strftime(buffer,sizeof(buffer),"%d-%m-%Y %H:%M:%S",timeinfo);
+    std::string str(buffer);
+
+    int seconds_to_run = 30; // default value
+    if (argc > 1) {
+        seconds_to_run = stoi(std::string(argv[1]));
+        std::cout << "running for " << seconds_to_run << " seconds" << std::endl;
+
+    }
+
+
+
+    // check to see if we get valid coordinates by checking GPS quality
+    while (numTrials > 0) {
+        try {
+            if (gps.getGPSData().gpsQuality != BoBRobotics::GPS::GPSQuality::INVALID) {
+                std::cout << " we have a valid measurement" << std::endl;
+                break;
+            }
+        } catch(...) {
+            std::cout << " measuring failed, trying again in 1 second " << "[" << maxTrials - numTrials << "/" << maxTrials << "]" << std::endl;
+            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+            numTrials--;
+        }
+    }
+    if (numTrials == 0) {
+        std::cout << " There is no valid gps measurement, please try waiting for the survey in to finish and restart the program " << std::endl;
+        exit(1);
```

Comment:
It's better practice to do ``return EXIT_FAILURE;`` instead because that will call your objects' destructors. ``exit()`` just kills your program without cleaning anything up.

Reply:
i.e. your motors...